<!DOCTYPE html>
<html>
  <head>
    <meta charset="utf-8" />
    <title>New test/hotspot/jtreg/vmTestbase/nsk/jdi/ExceptionRequest/exception/exception001.java</title>
    <link rel="stylesheet" href="../../../../../../../../style.css" />
  </head>
  <body>
    <pre>
  1 /*
  2  * Copyright (c) 2001, 2018, Oracle and/or its affiliates. All rights reserved.
  3  * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
  4  *
  5  * This code is free software; you can redistribute it and/or modify it
  6  * under the terms of the GNU General Public License version 2 only, as
  7  * published by the Free Software Foundation.
  8  *
  9  * This code is distributed in the hope that it will be useful, but WITHOUT
 10  * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
 11  * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
 12  * version 2 for more details (a copy is included in the LICENSE file that
 13  * accompanied this code).
 14  *
 15  * You should have received a copy of the GNU General Public License version
 16  * 2 along with this work; if not, write to the Free Software Foundation,
 17  * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
 18  *
 19  * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
 20  * or visit www.oracle.com if you need additional information or have any
 21  * questions.
 22  */
 23 
 24 package nsk.jdi.ExceptionRequest.exception;
 25 
 26 import nsk.share.*;
 27 import nsk.share.jpda.*;
 28 import nsk.share.jdi.*;
 29 
 30 import com.sun.jdi.*;
 31 import com.sun.jdi.event.*;
 32 import com.sun.jdi.request.*;
 33 
 34 import java.util.*;
 35 import java.io.*;
 36 
 37 /**
 38  * The test for the implementation of an object of the type     &lt;BR&gt;
 39  * ExceptionRequest.                                            &lt;BR&gt;
 40  *                                                              &lt;BR&gt;
 41  * The test checks that results of the method                   &lt;BR&gt;
 42  * &lt;code&gt;com.sun.jdi.ExceptionRequest.exception()&lt;/code&gt;        &lt;BR&gt;
 43  * complies with its spec.                                      &lt;BR&gt;
 44  * &lt;BR&gt;
 45  * The test checks if a boolean value, an argument of the method        &lt;BR&gt;
 46  *    EventRequestManager.createExceptionRequest(),                     &lt;BR&gt;
 47  * is equal to one returned by the method                               &lt;BR&gt;
 48  *    ExceptionRequest.exception()                                      &lt;BR&gt;
 49  * The cases to test include a ReferenceType                            &lt;BR&gt;
 50  * mirroring NullPointerException, and null.                            &lt;BR&gt;
 51  * &lt;BR&gt;
 52  * The test has three phases and works as follows.                      &lt;BR&gt;
 53  * &lt;BR&gt;
 54  * In first phase,                                                      &lt;BR&gt;
 55  * upon launching debuggee&#39;s VM which will be suspended,                &lt;BR&gt;
 56  * a debugger waits for the VMStartEvent within a predefined            &lt;BR&gt;
 57  * time interval. If no the VMStartEvent received, the test is FAILED.  &lt;BR&gt;
 58  * Upon getting the VMStartEvent, it makes the request for debuggee&#39;s   &lt;BR&gt;
 59  * ClassPrepareEvent with SUSPEND_EVENT_THREAD, resumes the VM,         &lt;BR&gt;
 60  * and waits for the event within the predefined time interval.         &lt;BR&gt;
 61  * If no the ClassPrepareEvent received, the test is FAILED.            &lt;BR&gt;
 62  * Upon getting the ClassPrepareEvent,                                  &lt;BR&gt;
 63  * the debugger sets up the breakpoint with SUSPEND_EVENT_THREAD        &lt;BR&gt;
 64  * within debuggee&#39;s special methodForCommunication().                  &lt;BR&gt;
 65  * &lt;BR&gt;
 66  * In second phase to check the assertion,                              &lt;BR&gt;
 67  * the debugger and the debuggee perform the following.                 &lt;BR&gt;
 68  * - The debugger                                                       &lt;BR&gt;
 69  *   - gets ReferenceType value mirroring NullPointerException, &lt;BR&gt;
 70  *   - sets up two ExceptionRequests with the ReferenceType and null    &lt;BR&gt;
 71  *     values of the exception arguments,                               &lt;BR&gt;
 72  *   - gets the values of the Request with the method                   &lt;BR&gt;
 73  *     StepRequest.exception(), and                                     &lt;BR&gt;
 74  *   - compares the values.                                             &lt;BR&gt;
 75  * &lt;BR&gt;
 76  * In third phase, at the end of the test, the debuggee changes         &lt;BR&gt;
 77  * the value of the &quot;instruction&quot; which the debugger and debuggee       &lt;BR&gt;
 78  * use to inform each other of needed actions,  and both end.           &lt;BR&gt;
 79  * &lt;BR&gt;
 80  */
 81 
 82 public class exception001 {
 83 
 84     //----------------------------------------------------- templete section
 85     static final int PASSED = 0;
 86     static final int FAILED = 2;
 87     static final int PASS_BASE = 95;
 88 
 89     //----------------------------------------------------- templete parameters
 90     static final String
 91     sHeader1 = &quot;\n==&gt; nsk/jdi/ExceptionRequest/exception/exception001 &quot;,
 92     sHeader2 = &quot;--&gt; debugger: &quot;,
 93     sHeader3 = &quot;##&gt; debugger: &quot;;
 94 
 95     //----------------------------------------------------- main method
 96 
 97     public static void main (String argv[]) {
 98 
 99         int result = run(argv, System.out);
100 
101         System.exit(result + PASS_BASE);
102     }
103 
104     public static int run (String argv[], PrintStream out) {
105 
106         int exitCode = new exception001().runThis(argv, out);
107 
108         if (exitCode != PASSED) {
109             System.out.println(&quot;TEST FAILED&quot;);
110         }
111         return testExitCode;
112     }
113 
114     //--------------------------------------------------   log procedures
115 
116     private static Log  logHandler;
117 
118     private static void log1(String message) {
119         logHandler.display(sHeader1 + message);
120     }
121     private static void log2(String message) {
122         logHandler.display(sHeader2 + message);
123     }
124     private static void log3(String message) {
125         logHandler.complain(sHeader3 + message);
126     }
127 
128     //  ************************************************    test parameters
129 
130     private String debuggeeName =
131         &quot;nsk.jdi.ExceptionRequest.exception.exception001a&quot;;
132 
133     //====================================================== test program
134     //------------------------------------------------------ common section
135 
136     static Debugee          debuggee;
137     static ArgumentHandler  argsHandler;
138 
139     static int waitTime;
140 
141     static VirtualMachine      vm            = null;
142     static EventRequestManager eventRManager = null;
143     static EventQueue          eventQueue    = null;
144     static EventSet            eventSet      = null;
145     static EventIterator       eventIterator = null;
146 
147     static ReferenceType       debuggeeClass = null;
148 
149     static int  testExitCode = PASSED;
150 
151 
152     //------------------------------------------------------ methods
153 
154     private int runThis (String argv[], PrintStream out) {
155 
156         argsHandler     = new ArgumentHandler(argv);
157         logHandler      = new Log(out, argsHandler);
158         Binder binder   = new Binder(argsHandler, logHandler);
159 
160         waitTime        = argsHandler.getWaitTime() * 60000;
161 
162         try {
163             log2(&quot;launching a debuggee :&quot;);
164             log2(&quot;       &quot; + debuggeeName);
165             if (argsHandler.verbose()) {
166                 debuggee = binder.bindToDebugee(debuggeeName + &quot; -vbs&quot;);
167             } else {
168                 debuggee = binder.bindToDebugee(debuggeeName);
169             }
170             if (debuggee == null) {
171                 log3(&quot;ERROR: no debuggee launched&quot;);
172                 return FAILED;
173             }
174             log2(&quot;debuggee launched&quot;);
175         } catch ( Exception e ) {
176             log3(&quot;ERROR: Exception : &quot; + e);
177             log2(&quot;       test cancelled&quot;);
178             return FAILED;
179         }
180 
181         debuggee.redirectOutput(logHandler);
182 
183         vm = debuggee.VM();
184 
185         eventQueue = vm.eventQueue();
186         if (eventQueue == null) {
187             log3(&quot;ERROR: eventQueue == null : TEST ABORTED&quot;);
188             vm.exit(PASS_BASE);
189             return FAILED;
190         }
191 
192         log2(&quot;invocation of the method runTest()&quot;);
193         switch (runTest()) {
194 
195             case 0 :  log2(&quot;test phase has finished normally&quot;);
196                       log2(&quot;   waiting for the debuggee to finish ...&quot;);
197                       debuggee.waitFor();
198 
199                       log2(&quot;......getting the debuggee&#39;s exit status&quot;);
200                       int status = debuggee.getStatus();
201                       if (status != PASS_BASE) {
202                           log3(&quot;ERROR: debuggee returned UNEXPECTED exit status: &quot; +
203                               status + &quot; != PASS_BASE&quot;);
204                           testExitCode = FAILED;
205                       } else {
206                           log2(&quot;......debuggee returned expected exit status: &quot; +
207                               status + &quot; == PASS_BASE&quot;);
208                       }
209                       break;
210 
211             default : log3(&quot;ERROR: runTest() returned unexpected value&quot;);
212 
213             case 1 :  log3(&quot;test phase has not finished normally: debuggee is still alive&quot;);
214                       log2(&quot;......forcing: vm.exit();&quot;);
215                       testExitCode = FAILED;
216                       try {
217                           vm.exit(PASS_BASE);
218                       } catch ( Exception e ) {
219                           log3(&quot;ERROR: Exception : &quot; + e);
220                       }
221                       break;
222 
223             case 2 :  log3(&quot;test cancelled due to VMDisconnectedException&quot;);
224                       log2(&quot;......trying: vm.process().destroy();&quot;);
225                       testExitCode = FAILED;
226                       try {
227                           Process vmProcess = vm.process();
228                           if (vmProcess != null) {
229                               vmProcess.destroy();
230                           }
231                       } catch ( Exception e ) {
232                           log3(&quot;ERROR: Exception : &quot; + e);
233                       }
234                       break;
235             }
236 
237         return testExitCode;
238     }
239 
240 
241    /*
242     * Return value: 0 - normal end of the test
243     *               1 - ubnormal end of the test
244     *               2 - VMDisconnectedException while test phase
245     */
246 
247     private int runTest() {
248 
249         try {
250             testRun();
251 
252             log2(&quot;waiting for VMDeathEvent&quot;);
253             getEventSet();
254             if (eventIterator.nextEvent() instanceof VMDeathEvent)
255                 return 0;
256 
257             log3(&quot;ERROR: last event is not the VMDeathEvent&quot;);
258             return 1;
259         } catch ( VMDisconnectedException e ) {
260             log3(&quot;ERROR: VMDisconnectedException : &quot; + e);
261             return 2;
262         } catch ( Exception e ) {
263             log3(&quot;ERROR: Exception : &quot; + e);
264             return 1;
265         }
266 
267     }
268 
269     private void testRun()
270                  throws JDITestRuntimeException, Exception {
271 
272         eventRManager = vm.eventRequestManager();
273 
274         ClassPrepareRequest cpRequest = eventRManager.createClassPrepareRequest();
275         cpRequest.setSuspendPolicy( EventRequest.SUSPEND_EVENT_THREAD);
276         cpRequest.addClassFilter(debuggeeName);
277 
278         cpRequest.enable();
279         vm.resume();
280         getEventSet();
281         cpRequest.disable();
282 
283         ClassPrepareEvent event = (ClassPrepareEvent) eventIterator.next();
284         debuggeeClass = event.referenceType();
285 
286         if (!debuggeeClass.name().equals(debuggeeName))
287            throw new JDITestRuntimeException(&quot;** Unexpected ClassName for ClassPrepareEvent **&quot;);
288 
289         log2(&quot;      received: ClassPrepareEvent for debuggeeClass&quot;);
290 
291         String bPointMethod = &quot;methodForCommunication&quot;;
292         String lineForComm  = &quot;lineForComm&quot;;
293 
294         ThreadReference   mainThread = debuggee.threadByNameOrThrow(&quot;main&quot;);
295 
296         BreakpointRequest bpRequest = settingBreakpoint(mainThread,
297                                              debuggeeClass,
298                                             bPointMethod, lineForComm, &quot;zero&quot;);
299         bpRequest.enable();
300         vm.resume();
301 
302     //------------------------------------------------------  testing section
303 
304         log1(&quot;     TESTING BEGINS&quot;);
305 
306         EventRequest  eventRequest1 = null;
307         EventRequest  eventRequest2 = null;
308 
309         String fieldName = &quot;obj&quot;;
310         Value value = null;
311 
312         ReferenceType refObject = null;
313         ReferenceType exception = null;
314 
315         for (int i = 0; ; i++) {
316 
317             breakpointForCommunication();
318 
319             int instruction = ((IntegerValue)
320                                (debuggeeClass.getValue(debuggeeClass.fieldByName(&quot;instruction&quot;)))).value();
321 
322             if (instruction == 0) {
323                 vm.resume();
324                 break;
325             }
326 
327             log1(&quot;:::::: case: # &quot; + i);
328 
329             //~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ variable part
330 
331             switch (i) {
332 
333               case 0:
334                      log2(&quot;......vm.resume();&quot;);
335                      vm.resume();
336 
337                      log2(&quot;......getting: refObject = debuggeeClass.getValue(debuggeeClass.fieldByName(fieldName));&quot;);
338                      value = debuggeeClass.getValue(debuggeeClass.fieldByName(fieldName));
339                      log2(&quot;      getting: refObject = (ReferenceType) value.type();&quot;);
340                      refObject = (ReferenceType) value.type();
341 
342                      log2(&quot;......setting up: eventRequest1 = setting25ExceptionRequest(refObject);&quot;);
343                      eventRequest1 = setting25ExceptionRequest(refObject);
344 
345                      log2(&quot;......getting: exception = ((ExceptionRequest) eventRequest1).exception();&quot;);
346                      exception = ((ExceptionRequest) eventRequest1).exception();
347 
348                      log2(&quot;      comparing ReferenceType objects&quot;);
349                      if ( !exception.equals(refObject) ) {
350                          testExitCode = FAILED;
351                          log3(&quot;ERROR: objects are not equal : &quot; + exception);
352                      }
353 
354                      log2(&quot;......setting up: eventRequest2 = setting25ExceptionRequest(null);&quot;);
355                      eventRequest2 = setting25ExceptionRequest(null);
356 
357                      log2(&quot;......getting: exception = ((ExceptionRequest) eventRequest2).exception();&quot;);
358                      exception = ((ExceptionRequest) eventRequest2).exception();
359 
360                      log2(&quot;      comparing to null&quot;);
361                      if ( exception != null ) {
362                          testExitCode = FAILED;
363                          log3(&quot;ERROR: exception is not the null : &quot; + exception);
364                      }
365 
366                       break;
367 
368               default:
369                       throw new JDITestRuntimeException(&quot;** default case 2 **&quot;);
370             }
371 
372 
373             //~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
374         }
375         log1(&quot;    TESTING ENDS&quot;);
376         return;
377     }
378 
379    /*
380     * private BreakpointRequest settingBreakpoint(ThreadReference, ReferenceType,
381     *                                             String, String, String)
382     *
383     * It sets up a breakpoint at given line number within a given method in a given class
384     * for a given thread.
385     *
386     * Return value: BreakpointRequest object  in case of success
387     *
388     * JDITestRuntimeException   in case of an Exception thrown within the method
389     */
390 
391     private BreakpointRequest settingBreakpoint ( ThreadReference thread,
392                                                   ReferenceType testedClass,
393                                                   String methodName,
394                                                   String bpLine,
395                                                   String property)
396             throws JDITestRuntimeException {
397 
398         log2(&quot;......setting up a breakpoint:&quot;);
399         log2(&quot;       thread: &quot; + thread + &quot;; class: &quot; + testedClass +
400                         &quot;; method: &quot; + methodName + &quot;; line: &quot; + bpLine);
401 
402         List              alllineLocations = null;
403         Location          lineLocation     = null;
404         BreakpointRequest breakpRequest    = null;
405 
406         try {
407             Method  method  = (Method) testedClass.methodsByName(methodName).get(0);
408 
409             alllineLocations = method.allLineLocations();
410 
411             int n =
412                 ( (IntegerValue) testedClass.getValue(testedClass.fieldByName(bpLine) ) ).value();
413             if (n &gt; alllineLocations.size()) {
414                 log3(&quot;ERROR:  TEST_ERROR_IN_settingBreakpoint(): number is out of bound of method&#39;s lines&quot;);
415             } else {
416                 lineLocation = (Location) alllineLocations.get(n);
417                 try {
418                     breakpRequest = eventRManager.createBreakpointRequest(lineLocation);
419                     breakpRequest.putProperty(&quot;number&quot;, property);
420                     breakpRequest.addThreadFilter(thread);
421                     breakpRequest.setSuspendPolicy( EventRequest.SUSPEND_EVENT_THREAD);
422                 } catch ( Exception e1 ) {
423                     log3(&quot;ERROR: inner Exception within settingBreakpoint() : &quot; + e1);
424                     breakpRequest    = null;
425                 }
426             }
427         } catch ( Exception e2 ) {
428             log3(&quot;ERROR: ATTENTION:  outer Exception within settingBreakpoint() : &quot; + e2);
429             breakpRequest    = null;
430         }
431 
432         if (breakpRequest == null) {
433             log2(&quot;      A BREAKPOINT HAS NOT BEEN SET UP&quot;);
434             throw new JDITestRuntimeException(&quot;**FAILURE to set up a breakpoint**&quot;);
435         }
436 
437         log2(&quot;      a breakpoint has been set up&quot;);
438         return breakpRequest;
439     }
440 
441 
442     private void getEventSet()
443                  throws JDITestRuntimeException {
444         try {
445 //            log2(&quot;       eventSet = eventQueue.remove(waitTime);&quot;);
446             eventSet = eventQueue.remove(waitTime);
447             if (eventSet == null) {
448                 throw new JDITestRuntimeException(&quot;** TIMEOUT while waiting for event **&quot;);
449             }
450 //            log2(&quot;       eventIterator = eventSet.eventIterator;&quot;);
451             eventIterator = eventSet.eventIterator();
452         } catch ( Exception e ) {
453             throw new JDITestRuntimeException(&quot;** EXCEPTION while waiting for event ** : &quot; + e);
454         }
455     }
456 
457 
458     private void breakpointForCommunication()
459                  throws JDITestRuntimeException {
460 
461         log2(&quot;breakpointForCommunication&quot;);
462         getEventSet();
463 
464         if (eventIterator.nextEvent() instanceof BreakpointEvent)
465             return;
466 
467         throw new JDITestRuntimeException(&quot;** event IS NOT a breakpoint **&quot;);
468     }
469 
470     // ============================== test&#39;s additional methods
471 
472     private ExceptionRequest setting25ExceptionRequest ( ReferenceType refType )
473             throws JDITestRuntimeException {
474         try {
475             log2(&quot;......setting up ExceptionRequest:&quot;);
476 
477             ExceptionRequest
478             excr = eventRManager.createExceptionRequest(refType, true, true);
479 
480             log2(&quot;      ExceptionRequest has been set up&quot;);
481             return excr;
482         } catch ( Exception e ) {
483             log3(&quot;ERROR: ATTENTION: Exception within settingExceptionRequest() : &quot; + e);
484             log3(&quot;       ExceptionRequest HAS NOT BEEN SET UP&quot;);
485             throw new JDITestRuntimeException(&quot;** FAILURE to set up ExceptionRequest **&quot;);
486         }
487     }
488 
489 }
    </pre>
  </body>
</html>